src/GlobImports/Exe.hs

```diff
@@ -78,80 +80,118 @@
 
 data AllModelsFile = AllModelsFile
     { amfModuleBase :: Module
     , amfModuleImports :: [Module]
     }
 
+printDebug :: Bool -> String -> IO ()
+printDebug enabled str = when enabled $ putStrLn ("[DEBUG] " ++ str)
+
 -- |
 --
 -- @since 0.1.0.0
 spliceImports
     :: Source
     -> SourceContents
     -> Destination
+    -> Maybe FilePath
+    -> String
+    -> [String]
+    -> Bool
     -> IO ()
-spliceImports (Source src) (SourceContents srcContents) (Destination dest) = do
-    let (dir, file) = splitFileName src
-    files <- filter (/= file) <$> getFilesRecursive dir
+spliceImports (Source src) (SourceContents srcContents) (Destination dest) msearchDir pat prefixes debug = do
+    let
+        (sourceDir, _file) = splitFileName src
+        searchDir = fromMaybe sourceDir msearchDir
+        excludePrefixFilter target = not $ any (`isPrefixOf` target) prefixes
+
+    printDebug debug $ "searching directory: " ++ searchDir
+    printDebug debug $ "searching with pattern: " ++ pat
+    printDebug debug $ "excluding file name: " ++ src
+    eitherFiles <- fmap (filter (essentiallyDistinct src)) <$> getFiles searchDir pat
+    files <- case eitherFiles of
+        Left e -> error e
+        Right f -> pure f
+    let
+        filteredFiles = filter excludePrefixFilter files
+    printDebug debug $ "including files:\n" ++ intercalate "\n|  " filteredFiles
     let
         input =
             AllModelsFile
                 { amfModuleBase =
                     fromJust $ pathToModule src
                 , amfModuleImports =
-                    mapMaybe pathToModule (fmap (dir </>) files)
+                    mapMaybe pathToModule (fmap (searchDir </>) filteredFiles)
                 }
         output =
             renderFile input srcContents
 
     writeFile dest output
+    where
+      essentiallyDistinct :: FilePath -> FilePath -> Bool
+      essentiallyDistinct l r = simplifyPath (Text.pack l) /= simplifyPath (Text.pack r)
+
+      simplifyPath :: Text -> Text
+      simplifyPath str =
+        let
+          intermediate = Text.replace "//" "/" str
+        in
+          if intermediate == str
+          then intermediate
+          else simplifyPath intermediate
 
 -- | Returns a list of relative paths to all files in the given directory.
-getFilesRecursive
+getFiles
     :: FilePath
+    -- ^ The glob pattern to filter with.
+    -> String
     -- ^ The directory to search.
-    -> IO [FilePath]
-getFilesRecursive baseDir = sort <$> go []
-  where
-    go :: FilePath -> IO [FilePath]
-    go dir = do
-      c <- map (dir </>) . filter (`notElem` [".", ".."]) <$> getDirectoryContents (baseDir </> dir)
-      dirs <- filterM (doesDirectoryExist . (baseDir </>)) c >>= mapM go
-      files <- filterM (doesFileExist . (baseDir </>)) c
-      return (files ++ concat dirs)
+    -> IO (Either String [FilePath])
+getFiles baseDir pat = do
+    (exitCode, out, err) <- readProcess $ proc "find" [baseDir, "-wholename", pat]
```

Comment:
We now have a runtime shell dependency on `find` with this change. How can we encode that in cabal? 

Also, are we confident that this works for different versions of `find`? Feel like differences in shell utilities are common with mac/linux/gnu/bsd/etc

Reply:
Good question, and I'm not at all sure it works on every possible version. In fact, I'd be surprised if it did.

The change is just for performance, so we could potentially make it optional and rely on the consumer to pass in the arguments.

Reply:
Or rely on the consumer to pass in the entire command, which I think would fairly eliminate the dependency as well and put the burden on the user to ensure that the tool is available.

`find` is part of the POSIX standard, at least, so most environments will have it.